### PR TITLE
Fix duplicate page listener error and other bugs.

### LINKED ISSUE #109: UI: Error adding tx and block notification listener: listener_already_exist
Error observed when using the app for the first time.
```
[ERR] UI: Error adding tx and block notification listener: listener_already_exist
```
Sorry I don't have any other context at this time not sure exactly when it happened.

### PR DESCRIPTION
**Pull request by ukane-philemon**
Close #109

- Fixes a bug that caused a child page to be initialized twice.
- Fixed a bug that prevents a new listener on the proposal details page (https://github.com/crypto-power/cryptopower/commit/ad7251862a3554b867615eceac62770323b0c6ee).

### CODE REVIEW

File: ui/page/dcrdex/dcrdex_page.go [hunk 1]
```diff
@@ -55,9 +55,9 @@ func (pg *DEXPage) OnNavigatedTo() {
 	if pg.CurrentPage() == nil {
 		// TODO: Handle pg.inited
 		pg.Display(NewDEXOnboarding(pg.Load))
+	} else {
+		pg.CurrentPage().OnNavigatedTo()
 	}
-
-	pg.CurrentPage().OnNavigatedTo()
```
Comment (ukane-philemon) [inline ui/page/dcrdex/dcrdex_page.go:60]: > Fixes a bug that caused a child page to be initialized twice.

I could have added this change in a separate commit but missed it.

File: ui/page/root/home_page.go [hunk 2]
```diff
@@ -136,6 +136,8 @@ func (hp *HomePage) OnNavigatedTo() {
 
 	if hp.CurrentPage() == nil {
 		hp.Display(NewOverviewPage(hp.Load))
+	} else {
+		hp.CurrentPage().OnNavigatedTo()
 	}
```
Comment (ukane-philemon) [inline ui/page/root/home_page.go:141]: Same as my comment above

File: ui/page/root/main_page.go [hunk 3]
```diff
@@ -301,6 +301,8 @@ func (mp *MainPage) OnNavigatedTo() {
 
 	if mp.CurrentPage() == nil {
 		mp.Display(info.NewInfoPage(mp.Load, redirect)) // TODO: Should pagestack have a start page?
+	} else {
+		mp.CurrentPage().OnNavigatedTo()
 	}
```
Comment (ukane-philemon) [inline ui/page/root/main_page.go:306]: and this too.

File: app/pagestack.go [hunk 4]
```diff
@@ -88,10 +88,11 @@ func (pageStack *PageStack) Pop() bool {
...
 	}
 
-	pageStack.pages = pageStack.pages[:l-1]
 	if l > 1 {
 		pageStack.pages[l-2].OnNavigatedTo() // get previous page ready for display
 	}
+
+	pageStack.pages = pageStack.pages[:l-1] // remove page.
 	return true
 }
```
Comment (ukane-philemon) [inline app/pagestack.go:97]: Found this too and wanted to add it here.

**Comment by dreacot** [comment]
the app panics when i navigate to the trade page

```
panic: send on closed channel

goroutine 231 [running]:
github.com/crypto-power/cryptopower/libwallet/ext.(*RateListener).Notify(...)
        /home/kennedy/Projects/src/github.com/cryptopower/cryptopower/libwallet/ext/rate_notification.go:16
github.com/crypto-power/cryptopower/libwallet/ext.(*CommonRateSource).notifyRateListeners(0xc003778480?)
        /home/kennedy/Projects/src/github.com/cryptopower/cryptopower/libwallet/ext/rate_source.go:471 +0x109
github.com/crypto-power/cryptopower/libwallet/ext.(*CommonRateSource).startWebsocket.func1()
        /home/kennedy/Projects/src/github.com/cryptopower/cryptopower/libwallet/ext/rate_source.go:384 +0xd2
created by github.com/crypto-power/cryptopower/libwallet/ext.(*CommonRateSource).startWebsocket
        /home/kennedy/Projects/src/github.com/cryptopower/cryptopower/libwallet/ext/rate_source.go:337 +0xae
exit status 2
```

**Comment by itswisdomagain** [comment]
@ukane-philemon If you don't mind, please hold off on this for a moment. I hadn't seen the issue #109 before but I'd noticed that the way listeners are used in the app currently is potentially buggy. I have a planned refactor on this coming soon.

File: libwallet/assets/btc/sync.go [hunk 5]
```diff
@@ -219,6 +219,10 @@ func (asset *Asset) handleNotifications() {
 
 notificationsLoop:
 	for {
+		if asset.syncCtx.Err() != nil {
+			break notificationsLoop // return early
+		}
```
Comment (itswisdomagain) [inline libwallet/assets/btc/sync.go:224]: This isn't a good idea. Sync doesn't stop exactly when the ctx is canceled. Allow notifications to be processed until sync stops completely. Pages that no longer care about sync ntfns can de-register their progress listeners.

**Review by itswisdomagain — CHANGES_REQUESTED** [review]
Let's revisit this PR and the original issue after #240 is merged.

File: libwallet/assets/btc/sync.go (hunk 5 above)
Comment (ukane-philemon) [reply libwallet/assets/btc/sync.go:224]: Okay. We exit below on context cancellation and I think we should give it more priority? All pages must have been closed before this context is canceled.

**Comment by ukane-philemon** [comment]
Since https://github.com/crypto-power/cryptopower/pull/240 is merged, I'll revisit and extract the other changes into a separate PR.

**Comment by ukane-philemon** [comment]
> Since #240 is merged, I'll revisit and extract the other changes into a separate PR.

See #250 (https://github.com/crypto-power/cryptopower/commit/5cc572e5e721e009d6b3e49f70ceae32cf2a16fc)

**Pull request closed without merging**